dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java

```diff
@@ -0,0 +1,374 @@
+        String candidateId = hash;
+
+        for (int i = 1; i <= MAX_ATTEMPTS; i++) {
+            if (!identifierFactory.isIdentifier(candidateId)) {
+                return candidateId;
+            }
+            candidateId = uuidSupplier.get();
+        }
+        Logger.warn(DeterministicIdentifierAPIImpl.class,String.format("Attempt to generate id has failed instead a non deterministic id (%s) was returned.",candidateId));
+        return candidateId;
+    }
+
+    @CloseDBIfOpened
+    private String bestEffortDeterministicContentId(final String hash) {
+
+        String candidateId = hash;
+
+        for (int i = 1; i <= MAX_ATTEMPTS; i++) {
+            if (!isContentTypeInode(candidateId)) {
+                return candidateId;
+            }
+            candidateId = uuidSupplier.get();
+        }
+        Logger.warn(DeterministicIdentifierAPIImpl.class,String.format("Attempt to generate content-type id has failed instead a non deterministic id (%s) was returned.",candidateId));
+        return candidateId;
+    }
+
+    /**
+     * Test if the calculated hash has already been used
+     * @param hash
+     * @return
+     */
+    private boolean isContentTypeInode(final String hash){
+
+            return new DotConnect()
+                    .setSQL("select count(*) as test from structure s join inode i on s.inode = i.inode where s.inode =?")
+                    .addParam(hash)
+                    .getInt("test")>0;
+
+    }
+
+    /**
+     * Entry point for (Contentlets, Host, Persona, Templates, Folders, FileAsset)
+     * @param asset
+     * @param parent
+     * @return
+     */
+    @Override
+    public String generateDeterministicIdBestEffort(final Versionable asset,
+            final Treeable parent) {
+        return enabled() ? bestEffortDeterministicId(hash(deterministicIdSeed(asset, parent)))
+                : uuidSupplier.get();
+    }
+
+    /**
+     * Entry point for Content-Types
+     * @param contentType CT to generate the id for
+     * @param contentTypeVarName sometimes the var-name isn't set on the contentType
+     * which is immutable in such cases we need to rely on this supplier.
+     * @return generated deterministic id
+     */
+    @Override
+    public String generateDeterministicIdBestEffort(final ContentType contentType,
+            final Supplier<String> contentTypeVarName) {
+        return enabled() ? bestEffortDeterministicContentId(
+                hash(deterministicIdSeed(contentType, contentTypeVarName)))
+                : uuidSupplier.get();
+    }
+
+    /**
+     * Given a Language this will evaluate the code and country code if any then generate a sha256 and finally will hash it out into a long val
+     * @param lang
+     * @return
+     */
+    @Override
+    public long generateDeterministicIdBestEffort(final Language lang) {
+        //You never know.. better to be safe than sorry
+        if("US".equals(lang.getCountryCode()) &&  "en".equals(lang.getLanguageCode())){
```

Comment:
@wezell not sure if this is all times true, we know some clients deletes all langs and start from scratch, so may be should be a config prop to turn on, turn off

Reply:
FYI:  clients could start dotcms with a diff default language than `en_US`

Reply:
Removing it